### ISSUE TITLE
Basic Controls

### ISSUE THREAD

**Issue by addonovan** [addonovan/freevicky#1]
At a minimum, there should be the following controls:
1. Basic Control
2. Label
3. Button

These are the most fundamental controls, and most other controls could be done with just a combination of these.

# 1.  Basic Control

A basic control is a control that's basic (who knew?). It is simply a texture handled by the ui system. It should have two states:

  1. Normal
  2. Hovered

If no hover texture is specified, then it should default to the normal texture.

# 2. Label

A label is a control that displays some text. Its text should be modifiable via a `set_text` call. Instead of a texture, this is a simply an `sf::Text` element handled by the ui system.

# 3. Button

A button is a mixture of the basic control and label. It should have all of the capabilities described for the basic control and label, as well as having a callback action for when it is pressed.

**Comment by addonovan** [comment]
In most frameworks, controls are handled through hierarchies or OOP design to prevent reimplementing functionality. We could do the same thing here, but it might be interesting to try a compositional approach instead.

A good example of this would be the button described above. Instead of inheriting from both the basic control and label classes, it could have both classes instantiated as its private members and delegate method calls to them.

```c++
class Button : public Control
{
  private:
    BasicControl m_basic;
    Label m_label;

  /* delegate control logic to both */
  public:
    void render( sf::RenderTarget& target )
    {
      m_basic.render( target );
      m_label.render( target );
    }

  /* delegate basic control logic */
  public:
    void set_texture( const std::string& texture_name ) { m_basic.set_texture( texture_name ); }

  /* delegate label logic */
  public:
    void set_text( const std::string& text ) { m_label.set_text( text ); }
    void set_font( const std::string& font_name ) { m_label.set_font( text ); }
}
```

Pros of composition:
* It leaves out inheritance, which may cause problems in the future
* It allows each control to implement only the methods it needs from its components

Cons:
* All delegate must be handled manually (C++ has no language-level support for this, unlike kotlin :heart:)
  * maybe this could be handled by some nasty preprocessor hack?
  * `DELEGATE( func_name, delegatee_name, return_type, (param_type, param_name)... )`
  * `DELEGATE( set_text, m_label, void, const std::string&, text )`

On the other hand, C++ does allow for multiple inheritance, unlike most languages that follow it. So we could potentially just use that feature; although, because not everything is planned out right now, it *may* lead to messy situations in the future.

**Comment by addonovan** [comment]
@TotalTechGeek
(I hope that works lol, never @'d anyone before)

**Comment by addonovan** [comment]
Implemented in 8a26b2480ab6a93218fffaa2b6ce776eb788e263. Quoting the commit message:

> the Control files contain the following:
> 1. Definition of Control interface (pure virtual class)
> 2. Label class
> 3. TextureControl class (this needs a better name...)
> 
> Label just draws text, while texture control just draws a texture.
> Currently, there is no positioning controls either.
> 
> The Button control is a composition of a Label with a TextureControl. I
> chose to perform composition instead of inheritance, as had inheritance
> been used here, then it would not have been able to add to a list,
> because it would be ambiguous which Control base should be used. And
> this way, Button now controls exactly what parts of the Label and
> TextureControl classes it implements, and can also control the
> positioning (to be done in future) of each independently.
> 
> The Container control is really just a container of Controls that itself
> implements the Control interface. At the moment, it just passes
> everything off to its children instead of processing anything.
